### DISCUSSION TITLE
What is the purpose of keyParts in unstable_cache since it cannot be revalidated ?

### DISCUSSION THREAD

**Discussion by anarkrypto** [vercel/next.js#63099] [Help] [answered]
### Summary

Example:
```js
const getCachedUser = unstable_cache(fetchUser, [`user-43`])
```

The first intuition I had about this key was that I could revalidate this with revalidateTags("user-43")
However that doesn't work, we must use

```js
const getCachedUser = unstable_cache(fetchUser, [`user-43`], {
  revalidate: false,
  tags: ["user-43"]
})
```

Is not clear why we must be repetitive

Current docs is not clear too
![image](https://github.com/vercel/next.js/assets/32111208/3be946db-dd3c-4915-bb95-355b355222b3)

**Comment by derekgygax** [comment]
I agree, I faced the same problem for a while. They do acknowledge this in the youtube video on caching in nextjs. However, the docs are not clear about this.

**Comment by anarkrypto** [comment]
I think cache without control is very dangerous.
Since we can only revalidate with "revalidate" and "tags" options, I don't see why we need keyParts.

It seems keyParts is a very counterintuitive and useless param. It can degrade dev experience, perhaps introducing some re-validation bugs because of this confusions.

**Comment by anarkrypto** [comment]
If keyParts is used to identify, why not use tags instead ?

Fetch cache works well without keyParts

**Comment by jeltehomminga** [comment]
yes also don't really get why we need both keyparts and tags 🤔

**Comment by romeovs** [comment, marked as answer]
Tags and cache keys are orthogonal concepts.

The key maps one-to-one to a cache entry. The tags don't have to. One tag can map to multiple cache entries (and so you can revalidate multiple cache entries with one call to `revalidateTag`).

For example, say you run a website that has a shop and a blog. There are two api's you talk to: shopify and wordpress.

Calls to the shop could look like:

```js
function getProductById(id) {
  return unstable_cache(() => shopify.getProduct(id), ["shopify", id], {
    tags: ["shopify", `shopify:${id}`],
  })
}
```

whereas the calls to wordpress could look like:
```js
function getBlogPostById(id) {
  return unstable_cache(() => wordpress.getPost(id), ["wordpress", id], {
    tags: ["wordpress" , "wordpress:${id}"],
  })
}
```

Now you add a global sale to your shop. You need to revalidate all products so the price can be updated. You can call `revalidateTag("shopify")`.

This will: 
1. allow you invalidate a group of tags (ie. all products in this case). This is not possible with just the key.
2. allow you to invalidate a specific group (ie. all the products, but not all the wordpress posts).
3. still allow you to invalidate a specific product/blog (ie. `revalidateTag("product:42")`)

**Reply by svdvs** [reply]
I still don't quite see how this addresses the original question. What is the keyParts array adding here? The revalidateTag call tag is working because you included the id in the tag, ie `shopify:${id}`, not because of the keyParts array (if my understanding is correct).

**Reply by Arinji2** [reply]
That's the thing, tags is for caching only. Keyparts is for identification 


A simple analogy, in react when we map over a list... We need to give each element a key right, couldn't react just do this from our code? No it can't you need to specify an id for it to identify correctly

Tags is for caching, keyparts is to identify multiple caches from one function.

**Reply by svdvs** [reply]
Thanks for the reply @Arinji2 , that does clear things up for me.

I do still think it's a little weird. Since it is an optional, it appears Next does have a way of identifying the function without needing the developer to specify. I think in the docs it even says that it's a combination of the function signature + parameters. There's lots of places where Next is already doing this sort of mapping automatically, like in the case of Request Memoization.  So while I still don't really get why it's there, you did help me understand what it is functionally doing.

**Reply by Arinji2** [reply]
Np, yes you are correct.. unstable cache can like do basic function analysis. Hence why keyparts is optional. You can either use that or just pass the id as args to the function signature so it can see the IDs changing. 

Basically you either help nextjs figure out the function is changing by passing ids as args.. or you just do it yourself by putting the IDs in keyparts


Unstable cache is still unstable (even though it's pretty much the most stable thing with nextjs app dir xD)  so the syntax might change. 

@svdvs

**Reply by joulev** [reply]
By the way, using string literals like `shopify` in the key part array is not necessary. They never change, so they never do anything. It's like

```tsx
<div key="shopify" />
```

the `key` here does nothing since it never changes.

That's why key parts are primarily useful only when it includes a variable, such as `id` in the example.

Note that I've observed that in most cases, you don't actually need key parts, and `unstable_cache` with keys can almost always be rewritten without keys. The code in the example could be rewritten as

```tsx
// I name it id1 and id2 to emphasise they are distinct,
// but in the code you would likely just name them all `id`
function getProductById(id1) {
  return unstable_cache((id2) => shopify.getProduct(id2), [], {
    tags: ["shopify", `shopify:${id1}`],
  })(id1);
}
```

(btw the code example is missing a `()` at the end of `unstable_cache`.)

**Reply by svdvs** [reply]
@Arinji2 I finally understand now, thank you! So it really is for instances where you aren't passing your params into the function, because Next can't index the function signature without the params. I remember reading something about this in the documentation but didn't understand it. I get it now.

**Reply by Arinji2** [reply]
Yes exactly, its an alternative method of passing props to make your function unique. Np, its a very normal thing for people to get confused with, once next 15 releases my changes to the unstable cache docs should make it much more simpler to understand :D

**Reply by lorikku** [reply]
I had another question about this. If I pass `keyParts`, will `unstable_cache` completely ignore any arguments/function identification and use `keyParts` exclusively? Or will it combine them?

**Reply by joulev** [reply]
@lorikku It’ll combine them. Essentially it internally creates a cache key that is generated from
* the passed function (`cb.toString()`)
* `keyParts.join(",")`
* arguments passed to the function (`JSON.stringify(args)`)

**Reply by lorikku** [reply]
That's what I assumed, thanks sm!

**Comment by Arinji2** [comment]
I recently made a PR to update the docs with a use case and a way to differentiate between when to use tags and when to use keyparts :D
https://github.com/vercel/next.js/pull/65942

**Comment by noc2spam** [comment]
This should really be added to 

https://nextjs.org/docs/app/api-reference/functions/unstable_cache

If it was there, I wouldn't have wasted so much time trying to figure out the right way. :(

**Reply by Arinji2** [reply]
A pr has been merged to canary regarding this, should be visible when next 15 is stable.
